### PR TITLE
added shopping cart entity

### PR DESCRIPTION
**Pull request by c-kalenyk**
Same problem with insertion by liquibase like in previous PR

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    public CartItemDto saveCartItem(Authentication authentication,
+                                    @RequestBody @Valid CreateCartItemRequestDto requestDto) {
+        UserResponseDto user = userService.findByEmail(authentication.getName());
+        return cartItemService.save(user.getId(), requestDto);
+    }
+
+    @Operation(summary = "Update cart item quantity", description = "Update "
+            + "cart item quantity by id")
+    @PutMapping("/items/{cartItemId}")
+    public CartItemDto updateCartItemQuantity(@PathVariable Long cartItemId,
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java]: better return whole ShoppingCart in response, cart item on its own will be hardly of use to the Front-end

File: src/main/java/mate/academy/bookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @Min(1)
+    private int quantity;
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
    @Positive
    private long bookId;
    @Positive
    private int quantity;

```

File: src/main/java/mate/academy/bookstore/dto/cartitem/UpdateCartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemRequestDto {
+    @Min(1)
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/dto/cartitem/UpdateCartItemRequestDto.java]: same

**Review by okuzan — APPROVED** [review]

**Comment by c-kalenyk** [comment]
Interesting that without Transactional annotations any attempts to save, get or update ends up with LazyInitializationException. But exclusively on final mapping stage, even though i'm now fetching all required data by custom queries.

File: src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+@RequiredArgsConstructor
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final UserService userService;
+
+    @Operation(summary = "Get all books", description = "Get a list of all available books")
+    @GetMapping
+    public ShoppingCartDto getAllCartItems(Authentication authentication) {
+        UserResponseDto user = userService.findByEmail(authentication.getName());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java]: don't need to call db, get userId from Authentication

File: src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return shoppingCartService.getCartItems(user.getId());
+    }
+
+    @Operation(summary = "Add a new cart item", description = "Add a new cart item "
+            + "to shopping cart")
+    @PostMapping
+    public ShoppingCartDto saveCartItem(Authentication authentication,
+                                    @RequestBody @Valid CreateCartItemRequestDto requestDto) {
+        UserResponseDto user = userService.findByEmail(authentication.getName());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java]: same here

File: src/main/java/mate/academy/bookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @Positive
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/cartitem/CreateCartItemRequestDto.java:9]:
```suggestion
    @Positive
    @NotNull
```

File: src/main/java/mate/academy/bookstore/dto/shoppingcart/ShoppingCartDto.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.bookstore.dto.shoppingcart;
+
+import java.util.Set;
+import lombok.Data;
+import mate.academy.bookstore.dto.cartitem.CartItemDto;
+
+@Data
+public class ShoppingCartDto {
+    private Long id;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/shoppingcart/ShoppingCartDto.java:9]:
```suggestion
    private Long id;
    Long userId;
```

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 8]
```diff
@@ -39,7 +39,7 @@ public class Book {
     private String description;
     private String coverImage;
     @Column(nullable = false)
-    private boolean isDeleted = false;
+    private boolean isDeleted;
     @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+@Entity
+@Getter
+@Setter
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+@SQLRestriction("is_deleted = FALSE")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    @MapsId
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart", fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 11]
```diff
@@ -40,7 +40,7 @@ public class User implements UserDetails {
     private String lastName;
     private String shippingAddress;
     @Column(nullable = false)
-    private boolean isDeleted = false;
+    private boolean isDeleted;
     @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/repository/book/BookRepository.java [hunk 12]
```diff
@@ -1,12 +1,17 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
     @Query("SELECT b FROM Book b JOIN b.categories a WHERE a.id = :categoryId")
     List<Book> findAllByCategoryId(Long categoryId);
+
+    @Query("SELECT b FROM Book b LEFT JOIN FETCH b.categories WHERE b.id = :id")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/book/BookRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "categories")
```

more readable

File: src/main/java/mate/academy/bookstore/repository/book/BookRepository.java [hunk 13]
```diff
@@ -1,12 +1,17 @@
...
+    Optional<Book> findById(@Param("id") Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/book/BookRepository.java]:
```suggestion
    Optional<Book> findById(Long id);

```

File: src/main/java/mate/academy/bookstore/repository/shoppingcart/ShoppingCartRepository.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("SELECT sc FROM ShoppingCart sc "
+            + "LEFT JOIN FETCH sc.user u "
+            + "LEFT JOIN FETCH sc.cartItems ci "
+            + "WHERE sc.id = :id")
+    Optional<ShoppingCart> findById(@Param("id") Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/shoppingcart/ShoppingCartRepository.java]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
    Optional<ShoppingCart> findById(Long id);
```

File: src/main/java/mate/academy/bookstore/service/UserService.java [hunk 15]
```diff
@@ -7,4 +7,6 @@
 public interface UserService {
     UserResponseDto register(UserRegistrationRequestDto requestDto)
             throws RegistrationException;
+
+    UserResponseDto findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/UserService.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Transactional
+    @Override
+    public ShoppingCartDto save(Long shoppingCartId, CreateCartItemRequestDto requestDto) {
+        Book book = bookRepository.findById(requestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find shopping cart by id: " + shoppingCartId));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]: wrong message

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        shoppingCart.getCartItems().add(cartItem);
+        return shoppingCartMapper.toDto(shoppingCartRepository.save(shoppingCart));
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto updateQuantity(Long id, UpdateCartItemRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find existing cart item by id: " + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId()).....
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        return shoppingCartMapper.toDto(shoppingCart);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        cartItemRepository.save(cartItem);
        return shoppingCartMapper.toDto(cart);
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto getCartItems(Long shoppingCartId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]: method is named getCartItems but you return ShoppingCartDto, fix it

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(shoppingCartId)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java:61]: If you try to get shop by user, you need to find shop cart by user id

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 21]
```diff
@@ -35,4 +36,11 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         user.setRoles(Set.of(role));
         return userMapper.toDto(userRepository.save(user));
     }
+
+    @Override
+    public UserResponseDto findByEmail(String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by c-kalenyk** [comment]
About shopping cart id and user id, i'm mapping it from user. There is only one cart per user anyway. So to simplify it a bit i can just use one id, no matter from which source of these two i will take it. If it's wrong - please, tell me

**Comment by Elena-Bruyako** [comment]
> About shopping cart id and user id, i'm mapping it from user. There is only one cart per user anyway. So to simplify it a bit i can just use one id, no matter from which source of these two i will take it. If it's wrong - please, tell me

let's leave as it is

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**